File: box-and-box/ports/go/govern.go
package main

// govern.go — three-modality decision (port of govern.mjs). Laws DB1–DB3.
// Precedence: ALETHIC ▸ DEONTIC ▸ AXIOLOGICAL.

import "sort"

// GovernOpts configures govern.
type GovernOpts struct {
	Req      Requirements
	Norms    []Norm
	Semiring string
}

// DeonticVeto records a deontically forbidden option.
type DeonticVeto struct {
	ID          string
	Status      string
	Overridable bool
	By          []string
	Overridden  []string
}

// AlethicVeto records an alethically infeasible option.
type AlethicVeto struct {
	ID       string
	Failures []Failure
}

// Escalation records a contrary-to-duty escalation.
type Escalation struct {
	Required      bool
	Repair        string
	Reason        string
	BlockedOption string
	BlockedBy     []Failure
}

// GovernRanking is a ranked admissible item.
type GovernRanking struct {
	ID     string
	Score  float64
	Status string
}

// GovernResult is the output of govern.
type GovernResult struct {
	Decision           string
	HasDecision        bool
	ForcedByObligation bool
	Escalation         *Escalation
	Margin             *float64
	Semiring           string
	Ranking            []GovernRanking
	DeonticallyVetoed  []DeonticVeto
	AlethicallyVetoed  []AlethicVeto
}

type govEval struct {
	id           string
	utility      float64
	ctx          map[string]interface{}
	value        Value
	feasible     bool
	feasFail     []Failure
	status       string
	overridden   []string
	contributors []Contributor
}

func hasContributor(e govEval, id string) bool {
	for _, c := range e.contributors {
		if c.ID == id {
			return true
		}
	}
	return false
}

func govern(options []Option, opts GovernOpts) GovernResult {
	semiringName := opts.Semiring
	if semiringName == "" {
		semiringName = "tropical"
	}
	s := semiring(semiringName)

	ev := make([]govEval, 0, len(options))
	for _, o := range options {
		feas := consume(o.Value, opts.Req)
		ctx := o.Ctx
		if ctx == nil {
			ctx = map[string]interface{}{}
		}
		v := resolve(adjudicateStatus(ctx, opts.Norms))
		ev = append(ev, govEval{
			id: o.ID, utility: o.util(s), ctx: ctx, value: o.Value,
			feasible: feas.Ok, feasFail: feas.Failures,
			status: v.Resolved, overridden: v.Overridden, contributors: v.Contributors,
		})
	}

	var alethicallyVetoed []AlethicVeto
	var survivors []govEval
	for _, e := range ev {
		if !e.feasible {
			alethicallyVetoed = append(alethicallyVetoed, AlethicVeto{ID: e.id, Failures: e.feasFail})
		} else {
			survivors = append(survivors, e)
		}
	}

	var deonticallyVetoed []DeonticVeto
	var admissible []govEval
	var obligatoryFeasible []govEval
	for _, e := range survivors {
		if e.status == FORBIDDEN {
			by := []string{}
			for _, c := range e.contributors {
				if c.Modality == "forbidden" {
					by = append(by, c.ID)
				}
			}
			deonticallyVetoed = append(deonticallyVetoed, DeonticVeto{ID: e.id, Status: e.status, Overridable: true, By: by, Overridden: e.overridden})
		}
		if e.status == OPTIONAL || e.status == OBLIGATORY {
			admissible = append(admissible, e)
			if e.status == OBLIGATORY {
				obligatoryFeasible = append(obligatoryFeasible, e)
			}
		}
	}

	var obligedButBlocked []govEval
	for _, e := range ev {
		if e.status == OBLIGATORY && !e.feasible {
			obligedButBlocked = append(obligedButBlocked, e)
		}
	}

	var escalation *Escalation
	if len(obligedButBlocked) > 0 && len(obligatoryFeasible) == 0 {
		blocked := obligedButBlocked[0]
		var nrm *Norm
		for i := range opts.Norms {
			if opts.Norms[i].Modality == "obligatory" && hasContributor(blocked, opts.Norms[i].ID) {
				nrm = &opts.Norms[i]
				break
			}
		}
		if nrm == nil {
			for i := range opts.Norms {
				if opts.Norms[i].Modality == "obligatory" {
					nrm = &opts.Norms[i]
					break
				}
			}
		}
		esc := escalate(nrm, blocked.ctx)
		repairID := "escalate-to-human"
		if esc.Repair != nil {
			repairID = esc.Repair.ID
		}
		escalation = &Escalation{Required: true, Repair: repairID, Reason: esc.Reason, BlockedOption: blocked.id, BlockedBy: blocked.feasFail}
	}

	if escalation == nil {
		for _, e := range survivors {
			if e.status == CONFLICT {
				escalation = &Escalation{Required: true, Repair: "escalate-to-human", Reason: "unresolved conflict on " + e.id, BlockedOption: e.id, BlockedBy: []Failure{}}
				break
			}
		}
	}

	pool := admissible
	if len(obligatoryFeasible) > 0 {
		pool = obligatoryFeasible
	}
	poolCopy := append([]govEval{}, pool...)
	sort.SliceStable(poolCopy, func(i, j int) bool { return poolCopy[i].utility > poolCopy[j].utility })

	res := GovernResult{Semiring: semiringName, DeonticallyVetoed: deonticallyVetoed, AlethicallyVetoed: alethicallyVetoed, Escalation: escalation}
	if escalation == nil && len(poolCopy) > 0 {
		res.Decision = poolCopy[0].id
		res.HasDecision = true
	}
	res.ForcedByObligation = res.HasDecision && len(obligatoryFeasible) > 0
	if len(poolCopy) > 1 {
		m := round3(poolCopy[0].utility - poolCopy[1].utility)
		res.Margin = &m
	}
	for _, e := range poolCopy {
		res.Ranking = append(res.Ranking, GovernRanking{ID: e.id, Score: round3(e.utility), Status: e.status})
	}
	return res
}

File: box-and-box/ports/rust/src/govern.rs
// govern.rs — three-modality decision (faithful port of govern.mjs). Laws DB1–DB3.
// Precedence: ALETHIC ▸ DEONTIC ▸ AXIOLOGICAL.

use crate::norm::{adjudicate_status, escalate, resolve, Ctx, Modality, Norm, Status};
use crate::score::Semiring;
use crate::value::{consume, round, Failure, Requirements, Value};

pub struct GOpt {
    pub id: String,
    pub value: Value,
    pub utility: Option<f64>,
    pub ctx: Ctx,
}

#[derive(Clone)]
pub struct DeonticVeto {
    pub id: String,
}

#[derive(Clone)]
pub struct AlethicVeto {
    pub id: String,
    pub failures: Vec<Failure>,
}

pub struct EscalationInfo {
    pub required: bool,
    pub repair: String,
    pub reason: String,
    pub blocked_option: String,
    pub blocked_by: Vec<Failure>,
}

pub struct RankEntry {
    pub id: String,
    pub score: f64,
    pub status: Status,
}

pub struct GovernResult {
    pub decision: Option<String>,
    pub forced_by_obligation: bool,
    pub escalation: Option<EscalationInfo>,
    pub margin: Option<f64>,
    pub ranking: Vec<RankEntry>,
    pub deontically_vetoed: Vec<DeonticVeto>,
    pub alethically_vetoed: Vec<AlethicVeto>,
}

struct Ev {
    id: String,
    utility: f64,
    ctx: Ctx,
    feasible: bool,
    feas_fail: Vec<Failure>,
    status: Status,
    #[allow(dead_code)]
    overridden: Vec<String>,
    contributors: Vec<crate::norm::Contributor>,
}

fn has(contributors: &[crate::norm::Contributor], id: &str) -> bool {
    contributors.iter().any(|c| c.id == id)
}

pub fn govern(options: &[GOpt], req: &Requirements, norms: &[Norm], s: Semiring) -> GovernResult {
    let ev: Vec<Ev> = options
        .iter()
        .map(|o| {
            let feas = consume(&o.value, req);
            let v = resolve(&adjudicate_status(&o.ctx, norms));
            Ev {
                id: o.id.clone(),
                utility: o.utility.unwrap_or_else(|| s.one()),
                ctx: o.ctx.clone(),
                feasible: feas.ok,
                feas_fail: feas.failures,
                status: v.resolved.unwrap_or(v.status),
                overridden: v.overridden,
                contributors: v.contributors,
            }
        })
        .collect();

    let alethically_vetoed: Vec<AlethicVeto> = ev
        .iter()
        .filter(|e| !e.feasible)
        .map(|e| AlethicVeto {
            id: e.id.clone(),
            failures: e.feas_fail.clone(),
        })
        .collect();

    let survivors: Vec<&Ev> = ev.iter().filter(|e| e.feasible).collect();

    let deontically_vetoed: Vec<DeonticVeto> = survivors
        .iter()
        .filter(|e| e.status == Status::Forbidden)
        .map(|e| DeonticVeto { id: e.id.clone() })
        .collect();

    let admissible: Vec<&&Ev> = survivors
        .iter()
        .filter(|e| e.status == Status::Optional || e.status == Status::Obligatory)
        .collect();
    let obligatory_feasible: Vec<&&Ev> = admissible
        .iter()
        .filter(|e| e.status == Status::Obligatory)
        .cloned()
        .collect();

    // contrary-to-duty
    let obliged_but_blocked: Vec<&Ev> = ev
        .iter()
        .filter(|e| e.status == Status::Obligatory && !e.feasible)
        .collect();

    let mut escalation: Option<EscalationInfo> = None;
    if !obliged_but_blocked.is_empty() && obligatory_feasible.is_empty() {
        let blocked = obliged_but_blocked[0];
        let nrm = norms
            .iter()
            .find(|n| n.modality == Modality::Obligatory && has(&blocked.contributors, &n.id))
            .or_else(|| norms.iter().find(|n| n.modality == Modality::Obligatory));
        let esc = escalate(nrm, &blocked.ctx);
        escalation = Some(EscalationInfo {
            required: true,
            repair: esc.repair.id.clone(),
            reason: esc.reason,
            blocked_option: blocked.id.clone(),
            blocked_by: blocked.feas_fail.clone(),
        });
    }
    // a surviving unresolved conflict also escalates
    let conflicted: Vec<&&Ev> = survivors
        .iter()
        .filter(|e| e.status == Status::Conflict)
        .collect();
    if escalation.is_none() && !conflicted.is_empty() {
        escalation = Some(EscalationInfo {
            required: true,
            repair: "escalate-to-human".into(),
            reason: format!("unresolved conflict on {}", conflicted[0].id),
            blocked_option: conflicted[0].id.clone(),
            blocked_by: Vec::new(),
        });
    }

    // selection — obligation forces the pool; otherwise rank all admissible
    let mut pool: Vec<&&Ev> = if !obligatory_feasible.is_empty() {
        obligatory_feasible.clone()
    } else {
        admissible.clone()
    };
    pool.sort_by(|a, b| b.utility.partial_cmp(&a.utility).unwrap());

    let chosen = if escalation.is_some() {
        None
    } else {
        pool.first().map(|e| e.id.clone())
    };
    let margin = if pool.len() > 1 {
        Some(round(pool[0].utility - pool[1].utility))
    } else {
        None
    };
    let ranking: Vec<RankEntry> = pool
        .iter()
        .map(|e| RankEntry {
            id: e.id.clone(),
            score: round(e.utility),
            status: e.status,
        })
        .collect();

    let forced_by_obligation = chosen.is_some() && !obligatory_feasible.is_empty();

    GovernResult {
        decision: chosen,
        forced_by_obligation,
        escalation,
        margin,
        ranking,
        deontically_vetoed,
        alethically_vetoed,
    }
}
